### PR TITLE
Fix EIP-7251 tests

### PR DESCRIPTION
**Pull request by ralexstokes**
supersedes #3648.

~spec fixes pulled out into #3657. will get that merged and rebase this to just the test fixes (may need to discuss some of them..)~

### CODE REVIEW

File: tests/core/pyspec/eth2spec/test/eip7251/block_processing/test_process_consolidation.py [hunk 1]
```diff
@@ -145,13 +144,11 @@ def test_consolidation_churn_limit_balance(spec, state):
...
     # This state has 256 validators each with 32 ETH in MINIMAL preset, 128 ETH consolidation churn
     consolidation_churn_limit = spec.get_consolidation_churn_limit(state)
-    # Set the consolidation balance to consume equal to churn limit
-    state.consolidation_balance_to_consume = consolidation_churn_limit
     current_epoch = spec.get_current_epoch(state)
 
     source_index = spec.get_active_validator_indices(state, current_epoch)[0]
     # Set source balance higher than consolidation churn limit
-    state.balances[source_index] = consolidation_churn_limit + 1
+    state.validators[source_index].effective_balance = 2 * consolidation_churn_limit
```
Comment (fradamt) [inline tests/core/pyspec/eth2spec/test/eip7251/block_processing/test_process_consolidation.py:151]:
```suggestion
    state.validators[source_index].effective_balance = consolidation_churn_limit + spec.EFFECTIVE_BALANCE_INCREMENT
```

I'd maybe do this instead, just to make it easier to interpret what happens. Here one sees the effective balance being 2x the churn limit, and it can be confusing why then `expected_exit_epoch` only gets a +1 instead of +2

**Comment by ralexstokes** [comment]
need to review but likely want to merge this in first: https://github.com/ralexstokes/consensus-specs/pull/3

**Review by hwwhww — APPROVED** [review]
well done 👍

note that some EIP-7251 results don't align with the testing goal/expections we had in the old tests. we'll need to have another scan of the missing edge cases.

File: tests/core/pyspec/eth2spec/test/phase0/epoch_processing/test_process_registry_updates.py [hunk 2]
```diff
@@ -205,16 +214,30 @@ def run_test_ejection_past_churn_limit(spec, state):
...
 
+    if is_post_eip7251(spec):
+        per_epoch_churn = spec.get_activation_exit_churn_limit(state)
+
+        def map_index_to_exit_epoch(i):
+            balance_so_far = i * spec.config.EJECTION_BALANCE
+            offset_epoch = balance_so_far // per_epoch_churn
+            if spec.config.EJECTION_BALANCE > per_epoch_churn - (balance_so_far % per_epoch_churn):
+                offset_epoch += 1
+            return expected_ejection_epoch + offset_epoch
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply tests/core/pyspec/eth2spec/test/phase0/epoch_processing/test_process_registry_updates.py:225]: I want to check the math here.. it works but im not sure this is precisely what we want

cc @fradamt

File: tests/core/pyspec/eth2spec/test/helpers/deposits.py [hunk 3]
```diff
@@ -253,6 +256,20 @@ def run_deposit_processing(spec, state, deposit, validator_index, valid=True, ef
...
+        updates_count = len(state.pending_balance_deposits)
+        if updates_count != 0:
+            assert updates_count == 1
+            pending_balance_deposit = state.pending_balance_deposits[0]
+            assert pending_balance_deposit.index == validator_index
+            assert pending_balance_deposit.amount == deposit.data.amount
+            spec.increase_balance(state, validator_index, deposit.data.amount)
+            if not is_top_up:
+                # run effective balance update for new validators
+                spec.process_effective_balance_updates(state)
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply tests/core/pyspec/eth2spec/test/helpers/deposits.py:271]: yeah, may end up changing this some, @fradamt has a PR linked above that may be a better way to cleanly handle the new vs old behavior

File: tests/core/pyspec/eth2spec/test/phase0/epoch_processing/test_process_registry_updates.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fradamt) [reply tests/core/pyspec/eth2spec/test/phase0/epoch_processing/test_process_registry_updates.py:225]: I think this currently breaks in the case when `spec.config.EJECTION_BALANCE == per_epoch_churn - (balance_so_far % per_epoch_churn)` and `offset_epoch > 0`, i.e., when the current exit exactly fills up the churn in an epoch after `expected_ejection_epoch`. Here, you wouldn't increase the `offset_epoch`, which makes sense in principle, but doesn't fit the current behavior, where the exit epoch actually jumps to the next one in that case. 
It should be correct with the change proposed here: https://github.com/ethereum/consensus-specs/pull/3668/files#r1562268491

[the remark this answers is not in the document]
Comment (fradamt) [reply tests/core/pyspec/eth2spec/test/phase0/epoch_processing/test_process_registry_updates.py:225]: Thought it was not going to work in an edge case because of a somewhat strange behavior of the current `compute_exit_epoch_and_update_churn` function, but on a deeper look I think it's always fine as long `spec.config.EJECTION_BALANCE` is not a multiple of the churn limit (which will always be the case unless we increase the ejection balance by a lot)

File: tests/core/pyspec/eth2spec/test/helpers/withdrawals.py [hunk 4]
```diff
@@ -1,4 +1,13 @@
 import random
+from eth2spec.test.helpers.forks import is_post_eip7251
+
+
+def get_expected_withdrawals(spec, state):
+    if is_post_eip7251(spec):
+        withdrawals, _ = spec.get_expected_withdrawals(state)
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply tests/core/pyspec/eth2spec/test/helpers/withdrawals.py:7]: yeah, if we look at how it is used, we could instead just `pop()` from the list and leave this function alone

File: tests/core/pyspec/eth2spec/test/capella/sanity/test_blocks.py [hunk 5]
```diff
@@ -357,7 +359,11 @@ def test_top_up_and_partial_withdrawable_validator(spec, state):
 
     signed_block = state_transition_and_sign_block(spec, state, block)
 
-    yield 'blocks', [signed_block]
+    # ensure we go through an epoch transition, to account for post-EIP-7251 behavior
+    block_in_next_epoch = build_empty_block(spec, state, slot=state.slot + spec.SLOTS_PER_EPOCH)
+    signed_block_in_next_epoch = state_transition_and_sign_block(spec, state, block_in_next_epoch)
+
+    yield 'blocks', [signed_block, signed_block_in_next_epoch]
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply tests/core/pyspec/eth2spec/test/capella/sanity/test_blocks.py:366]: I think its fine as extending the blocks is backwards-compatible with the `capella` and `deneb` behavior

File: tests/core/pyspec/eth2spec/test/helpers/deposits.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ralexstokes) [reply tests/core/pyspec/eth2spec/test/helpers/deposits.py]: update: merged in fradamt's PR that handles this a bit more directly
